pkg/resource/plugin/analyzer_config.go

```diff
@@ -0,0 +1,280 @@
+			// Root errors are prefixed with "(root):" (e.g. "(root): foo is required"),
+			// but that's just noise for our purposes, so we trim it from the message.
+			msg := strings.TrimPrefix(err.String(), "(root): ")
+			errors = append(errors, msg)
+		}
+	}
+	return errors, nil
+}
+
+func convertSchema(schema AnalyzerPolicyConfigSchema) JSONSchema {
+	result := JSONSchema{}
+	result["type"] = "object"
+	if len(schema.Properties) > 0 {
+		result["properties"] = schema.Properties
+	}
+	if len(schema.Required) > 0 {
+		result["required"] = schema.Required
+	}
+	return result
+}
+
+// CreateConfigWithDefaults returns a new map filled-in with defaults from the policy metadata.
+func CreateConfigWithDefaults(policies []AnalyzerPolicyInfo) (map[string]AnalyzerPolicyConfig, error) {
+	result := make(map[string]AnalyzerPolicyConfig)
+
+	// Prepare the resulting config with all defaults from the policy metadata.
+	for _, policy := range policies {
+		var props map[string]interface{}
+
+		// Set default values from the schema.
+		if policy.ConfigSchema != nil {
+			props = make(map[string]interface{})
+			for k, v := range policy.ConfigSchema.Properties {
+				if val, ok := v["default"]; ok {
+					props[k] = val
+				}
+			}
+		}
+
+		result[policy.Name] = AnalyzerPolicyConfig{
+			EnforcementLevel: policy.EnforcementLevel,
+			Properties:       props,
+		}
+	}
+
+	return result, nil
+}
+
+// ReconcilePolicyPackConfig takes metadata about each policy containing default values and config schema, and
+// reconciles this with the given config to produce a new config that has all default values filled-in and then sets
+// configured values.
+func ReconcilePolicyPackConfig(
+	policies []AnalyzerPolicyInfo, config map[string]AnalyzerPolicyConfig) (map[string]AnalyzerPolicyConfig, error) {
+	// Prepare the resulting config with all defaults from the policy metadata.
+	result, err := CreateConfigWithDefaults(policies)
+	contract.Assertf(result != nil, "result != nil")
+	if err != nil {
+		return nil, err
+	}
+
+	// Next, if the given config has "all" and an enforcement level, set it for all
+	// policies.
+	if config != nil {
+		if all, hasAll := config["all"]; hasAll && all.EnforcementLevel.IsValid() {
+			for _, v := range result {
+				v.EnforcementLevel = all.EnforcementLevel
+			}
+		}
+	}
+
+	// Next, loop through the given config, and set values.
+	for policy, givenConfig := range config {
+		resultConfig, hasResultConfig := result[policy]
+		if !hasResultConfig {
+			continue
+		}
+
+		if givenConfig.EnforcementLevel != "" {
+			resultConfig.EnforcementLevel = givenConfig.EnforcementLevel
+		}
+		if len(givenConfig.Properties) > 0 && resultConfig.Properties == nil {
+			resultConfig.Properties = make(map[string]interface{})
+		}
+		for k, v := range givenConfig.Properties {
+			resultConfig.Properties[k] = v
+		}
+	}
+
+	// Validate the resulting config.
+	if config != nil {
+		if err := ValidatePolicyPackConfig(policies, result); err != nil {
```

Comment:
Validation errors will be things like if a required config value is missing, or if the specified value is the incorrect type, etc.

But I actually don't think this is the right long-term place to be outputting such errors. This is what output currently looks like; it's just an error that stops the preview/update:

```bash
$ pulumi preview --policy-pack ~/path/to/policypack --policy-pack-config ~/path/to/policypack/config.json
Previewing update (dev):
 
Permalink: https://app.pulumi.com/<org>/<proj>/dev/previews/ae0e97ff-cd01-4071-9f71-746b1dca0695
error: error reconciling configuration for analyzer "aws-typescript" from path "/Users/user/path/to/policypack: 1 error occurred:
	* Policy "s3-no-public-read": foo is required
```

Would it be better if these were outputted similar to other policy diagnostic events?

Reply:
I think yes it probably should be similar to a diagnostic event, but maybe a new kind since this wont be tied to a particular resource 🤔 ... if its not a diagnostic event the service wont know about it!

I am also wondering if we should be validating the nil config. What happens if the config has an empty required field?

Reply:
Instead of returning these as a single combined error, I added a commit that emits these as individual diagnostic messages. This allows these errors to be seen in the Pulumi Console.

<img width="1442" alt="Screen Shot 2020-03-06 at 9 11 20 AM" src="https://user-images.githubusercontent.com/710598/76105759-75850980-5f8a-11ea-8780-30bc7e8e5b5d.png">

I also fixed it so all validation errors for all policy packs are returned, instead of returning early when the first policy pack had validation errors. 

We should still consider adding a new diagnostic event for these, which would allow improved grouping and display, but that requires changes to both the CLI and service and can be done as a subsequent improvement.

Reply:
> I am also wondering if we should be validating the nil config. What happens if the config has an empty required field?

Yep, fixed.